### ISSUE TITLE
Inverse rendering: Optimization assigns nan to vertices with flipped triangles

### ISSUE THREAD

**Issue by garrido-pablo** [mitsuba-renderer/mitsuba2#375]
*Please add one of the following label to your issue's title, and delete this section:*

- [🐛 bug report]

## Summary

I'm trying to optimize vertex positions via inverse rendering and I notice that as soon as the triangle face get flipped or becomes 'bad' their corresponding vertices are assigned nan.
To enable vertex differentiation, I had to do some changes to the source code, as described in https://github.com/mitsuba-renderer/mitsuba2/issues/305

Vertex optimization is not very stable for sure due to noisy renderings, but I don't know what kind of vertex update procedure is causing this or why that happens.

## System configuration

*For bug report, please enter information regarding your system configuration<- [remove this]*

- Platform: Windows 10 Enterprise / Cuda 10.2 / Optix 6.5
- Compiler: MSVC 19.25.28614.0
- Python version: 3.8.2
- Pytorch version: 1.4.0
- Mitsuba 2 version: 2.2.1 (b87c11a; but edited according to https://github.com/mitsuba-renderer/mitsuba2/issues/305)
- Compiled variants:
    * `scalar_rgb`
    * `gpu_autodiff_rgb`

## Description

Vertices have nan values as soon as triangles get flipped during vertex optimization. After that, the optimization will eventually start diverging, causing more nan values (which is not surprising).
Notes:
* I use render_torch() and pytorch to generate and backpropagate gradients.
* Texture optimization has NO issues on my current version.

This issue follows up issue reported on this link: https://github.com/mitsuba-renderer/mitsuba2/issues/305

## Steps to reproduce

1. Load scene
2. Get scene parameters
3. Get torch parameters from scene parameters
4. Setup pytorch optimizer (Adam)
5. Iterate N times
    * render scene using render_torch()
    * compute MSE loss
    * back-propagate gradients
    * update parameters (gradient descent step)

Unfortunately I can't share the code.

It is clear that vertex optimization is unstable due to noise and can cause triangle flipping, but vertices shouldn't be assigned nan when their corresponding triangle faces are flipped. Note that I also used Laplacian-based regularizers to avoid flipping but it doesn't prevent that from happening. Still, vertices should not be assigned nan when triangles are flipped.

Could you please point me to the source code where the nan vertex assignment might be happening? Or do you think this might be related to other issues related to vertex differentiation?

Thank you

**Comment by binorchen** [comment]
Hi @garrido-pablo.
I have also encountered the NaN problem even a little bit different from yours.
Would you like to try my [compromise solution](https://github.com/mitsuba-renderer/mitsuba2/issues/371#issuecomment-759728391)?  I also want to know whether this method works on other NaN problem or not.
Good luck.

**Comment by garrido-pablo** [comment]
Thanks for sharing your solution, @binorchen!

In the end, I checked the validity of the gradients after doing the backward pass. I simply kept valid gradients and ignored the rest (NaNs), either by assigning zeros or by restoring the value to the previous valid accumulated gradients - either option was selected depending on whether my loss function had a single or multiple constraints.
I found this solution much cleaner than altering the parameters' values directly.

I suspect the error comes from unstable floating-point operations while doing automatic differentiation. Maybe there are some missing sanity checks, which might be hard to spot in the source code.

I hope this helps others with similar issues.
